### PR TITLE
Shape Up #1 - Fourth prototype: workflow metadata and visualizer

### PR DESCRIPTION
**Pull request by lorenzo-cavazzi**
_work in progress, more info soon_

renku-python reference: https://github.com/SwissDataScienceCenter/renku-python/pull/3135
/deploy renku-core=workflow-ui-feature-branch #persist

### CODE REVIEW

**Review by andre-code — CHANGES_REQUESTED** [review]
Everything works as expected and the metadata is displayed correctly 🎉.
I leave a couple of questions to understand better the implementation.
And my only suggestion is to move the new components created in `client/src/workflows/Workflows.present.tsx` to separate files maybe in `/client/src/workflows/components`, as that file could grow a lot if we keep those auxiliary components there.

File: client/src/workflows/Workflows.state.ts [hunk 1]
```diff
@@ -59,6 +65,22 @@ function adjustWorkflowsList(workflowsList: Array<Record<string, any>>, fullPath
...
 
+/**
+ * Enrich workflow details by adding or modifying required by the UI
+ * @param workflowDetails - workflow details object as returned by the API
+ * @returns object containing enhanced workflow details
+ */
+function adjustWorkflowDetails(workflowDetails: Record<string, any>, fullPath: string) {
+  return {
+    ...workflowDetails,
+    latestUrl: workflowDetails.latest === workflowDetails.id ?
```
[the remark this answers is not in the document]
Comment (lorenzo-cavazzi) [reply client/src/workflows/Workflows.state.ts:76]: It's possible there is a newer version of the workflow. In that case, we show a warning and a link to the newest version. To minimize the computation, I compare here `latest` with `id` and add a new property with the URL to the newest version when it's available -- this is done only once just after fetching the workflow details data, and the presentational component can there just test for `workflow.details.latestUrl`.

File: client/src/utils/components/entities/Duration.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  duration: number | null;
+  workflowId: string;
+}
+
+function EntityDuration({
+  duration, workflowId
+}: EntityDurationProps) {
+  if (duration == null)
+    return null;
+  const elemId = `duration-${workflowId}`;
```
[the remark this answers is not in the document]
Comment (lorenzo-cavazzi) [reply client/src/utils/components/entities/Duration.tsx:35]: Ah, good catch!
Yes, the id would be repeated. I've just added a change to introduce a new `uniqueId` that changes based on the parent component. That is now used for the element ids

**Comment by lorenzo-cavazzi** [comment]
Thanks for the review! I changed the ids to be unique.

Would you split each workflow element into a single component file? I personally find it a bit dispersive since it's unlikely those components would be reused, while the (theoretically) generic components like the `TreeXyz` are already split apart.
If you think it's for the better, I'm ok with that. I feel it would be better to do it together with the bigger re-style and re-organization of the content you are going to do in for the 5th prototype since some components might be merged together and others split apart. Is this ok with you?

File: client/src/workflows/Workflows.state.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (andre-code) [reply client/src/workflows/Workflows.state.ts:77]: Ok, thanks for clarifying it

**Comment by andre-code** [comment]
Yes, we can include those changes in the next PR.

**Review by andre-code — APPROVED** [review]
🚀

**Pull request merged**